csharp/src/Google.Protobuf/WritingPrimitives.cs

```diff
@@ -0,0 +1,568 @@
+                value >>= 7;
+            }
+            if (state.position < state.limit)
+            {
+                buffer[state.position++] = (byte)value;
+            }
+            else
+            {
+                WriteRawByte(ref buffer, ref state, (byte)value);
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void WriteRawVarint64(ref Span<byte> buffer, ref WriterInternalState state, ulong value)
+        {
+            while (value > 127 && state.position < state.limit)
+            {
+                buffer[state.position++] = (byte)((value & 0x7F) | 0x80);
+                value >>= 7;
+            }
+            while (value > 127)
+            {
+                WriteRawByte(ref buffer, ref state, (byte)((value & 0x7F) | 0x80));
+                value >>= 7;
+            }
+            if (state.position < state.limit)
+            {
+                buffer[state.position++] = (byte)value;
+            }
+            else
+            {
+                WriteRawByte(ref buffer, ref state, (byte)value);
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void WriteRawLittleEndian32(ref Span<byte> buffer, ref WriterInternalState state, uint value)
+        {
+            if (state.position + 4 > state.limit)
+            {
+                WriteRawByte(ref buffer, ref state, (byte)value);
+                WriteRawByte(ref buffer, ref state, (byte)(value >> 8));
+                WriteRawByte(ref buffer, ref state, (byte)(value >> 16));
+                WriteRawByte(ref buffer, ref state, (byte)(value >> 24));
+            }
+            else
+            {
+                buffer[state.position++] = ((byte)value);
+                buffer[state.position++] = ((byte)(value >> 8));
+                buffer[state.position++] = ((byte)(value >> 16));
+                buffer[state.position++] = ((byte)(value >> 24));
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void WriteRawLittleEndian64(ref Span<byte> buffer, ref WriterInternalState state, ulong value)
+        {
+            if (state.position + 8 > state.limit)
+            {
+                WriteRawByte(ref buffer, ref state, (byte)value);
+                WriteRawByte(ref buffer, ref state, (byte)(value >> 8));
+                WriteRawByte(ref buffer, ref state, (byte)(value >> 16));
+                WriteRawByte(ref buffer, ref state, (byte)(value >> 24));
+                WriteRawByte(ref buffer, ref state, (byte)(value >> 32));
+                WriteRawByte(ref buffer, ref state, (byte)(value >> 40));
+                WriteRawByte(ref buffer, ref state, (byte)(value >> 48));
+                WriteRawByte(ref buffer, ref state, (byte)(value >> 56));
+            }
+            else
+            {
+                buffer[state.position++] = ((byte)value);
```

Comment:
BinaryPrimitives.WriteUInt64LittleEndian?

Reply:
benchmarks show it's actually slower, so added a comment.